File: central_rn/src/client/BlerpcClient.ts
import {
  ContainerSplitter,
  ContainerAssembler,
  Container,
  ContainerType,
  ControlCmd,
  CommandPacket,
  CommandType,
  makeTimeoutRequest,
  makeCapabilitiesRequest,
  makeStreamEndC2P,
  makeKeyExchange,
  centralPerformKeyExchange,
  BlerpcCryptoSession,
  CAPABILITY_FLAG_ENCRYPTION_SUPPORTED,
  BLERPC_ERROR_RESPONSE_TOO_LARGE,
} from '@blerpc/protocol-rn';
import { BleTransport, ScannedDevice } from '../ble/BleTransport';
import { GeneratedClient } from './GeneratedClient';

export class PayloadTooLargeError extends Error {
  constructor(
    public actual: number,
    public limit: number,
  ) {
    super(`Request payload (${actual} bytes) exceeds peripheral limit (${limit} bytes)`);
  }
}

export class ResponseTooLargeError extends Error {
  constructor(message: string) {
    super(`ResponseTooLargeError: ${message}`);
  }
}

export class PeripheralErrorException extends Error {
  constructor(public errorCode: number) {
    super(`PeripheralErrorException: 0x${errorCode.toString(16).padStart(2, '0')}`);
  }
}

export class ProtocolException extends Error {
  constructor(message: string) {
    super(`ProtocolException: ${message}`);
  }
}

export class BlerpcClient extends GeneratedClient {
  readonly transport = new BleTransport();
  private readonly requireEncryption: boolean;

  private _splitter: ContainerSplitter | null = null;
  private readonly _assembler = new ContainerAssembler();
  private _timeout = 100;
  private _maxRequestPayloadSize: number | null = null;

  // Encryption state
  private _session: BlerpcCryptoSession | null = null;

  constructor(requireEncryption = true) {
    super();
    this.requireEncryption = requireEncryption;
  }

  get mtu(): number {
    return this.transport.mtu;
  }

  get isEncrypted(): boolean {
    return this._session !== null;
  }

  private _readTimeout(firstRead: boolean): number {
    if (!firstRead) return this._timeout;
    return this._timeout > 2000 ? this._timeout : 2000;
  }

  private _handleControlError(container: Container): void {
    if (container.controlCmd === ControlCmd.ERROR && container.payload.length > 0) {
      const errorCode = container.payload[0];
      if (errorCode === BLERPC_ERROR_RESPONSE_TOO_LARGE) {
        throw new ResponseTooLargeError("Response exceeds peripheral's max_response_payload_size");
      }
      throw new PeripheralErrorException(errorCode);
    }
  }

  async scan(timeout?: number): Promise<ScannedDevice[]> {
    return this.transport.scan(timeout ?? 5000);
  }

  async connect(device: ScannedDevice): Promise<void> {
    await this.transport.connect(device);
    this._splitter = new ContainerSplitter(this.transport.mtu);

    try {
      await this._requestTimeout();
    } catch {
      console.log('Peripheral did not respond to timeout request, using default');
    }
    try {
      await this._requestCapabilities();
    } catch {
      console.log('Peripheral did not respond to capabilities request');
    }

    if (this.requireEncryption && this._session === null) {
      throw new Error(
        'Encryption required but key exchange was not completed. ' +
          'The peripheral may not support encryption or a MitM may ' +
          'have stripped the encryption capability flag.',
      );
    }
  }

  private async _requestTimeout(): Promise<void> {
    const s = this._splitter!;
    const tid = s.nextTransactionId();
    const req = makeTimeoutRequest(tid);
    await this.transport.write(req.serialize());
    const data = await this.transport.readNotify(1000);
    const resp = Container.deserialize(data);
    if (
      resp.containerType === ContainerType.CONTROL &&
      resp.controlCmd === ControlCmd.TIMEOUT &&
      resp.payload.length === 2
    ) {
      const bd = new DataView(
        resp.payload.buffer,
        resp.payload.byteOffset,
        resp.payload.byteLength,
      );
      const ms = bd.getUint16(0, true);
      this._timeout = ms;
      console.log(`Peripheral timeout: ${ms}ms`);
    }
  }

  private async _requestCapabilities(): Promise<void> {
    const s = this._splitter!;
    const tid = s.nextTransactionId();
    const req = makeCapabilitiesRequest(tid);
    await this.transport.write(req.serialize());
    const data = await this.transport.readNotify(1000);
    const resp = Container.deserialize(data);
    if (
      resp.containerType === ContainerType.CONTROL &&
      resp.controlCmd === ControlCmd.CAPABILITIES &&
      resp.payload.length >= 6
    ) {
      const bd = new DataView(
        resp.payload.buffer,
        resp.payload.byteOffset,
        resp.payload.byteLength,
      );
      const maxReq = bd.getUint16(0, true);
      const flags = bd.getUint16(4, true);
      this._maxRequestPayloadSize = maxReq;
      console.log(
        `Capabilities: max_req=${maxReq}, flags=0x${flags.toString(16).padStart(4, '0')}`,
      );

      if (flags & CAPABILITY_FLAG_ENCRYPTION_SUPPORTED) {
        await this._performKeyExchange();
      }
    }
  }

  private async _performKeyExchange(): Promise<void> {
    const s = this._splitter!;

    try {
      this._session = await centralPerformKeyExchange({
        send: async (payload: Uint8Array) => {
          const tid = s.nextTransactionId();
          const req = makeKeyExchange(tid, payload);
          await this.transport.write(req.serialize());
        },
        receive: async () => {
          const data = await this.transport.readNotify(2000);
          const resp = Container.deserialize(data);
          if (
            resp.containerType !== ContainerType.CONTROL ||
            resp.controlCmd !== ControlCmd.KEY_EXCHANGE
          ) {
            throw new Error('Expected KEY_EXCHANGE response, got something else');
          }
          return resp.payload;
        },
      });
      console.log('E2E encryption established');
    } catch (e) {
      console.log('Key exchange failed:', e);
      if (this.requireEncryption) throw e;
    }
  }

  private _encryptPayload(payload: Uint8Array): Uint8Array {
    if (this._session === null) {
      if (this.requireEncryption) {
        throw new Error('Encryption required but no session established');
      }
      return payload;
    }
    return this._session.encrypt(payload);
  }

  private _decryptPayload(payload: Uint8Array): Uint8Array {
    if (this._session === null) {
      if (this.requireEncryption) {
        throw new Error('Encryption required but no session established');
      }
      return payload;
    }
    return this._session.decrypt(payload);
  }

  protected async call(cmdName: string, requestData: Uint8Array): Promise<Uint8Array> {
    const s =
      this._splitter ??
      (() => {
        throw new Error('Not connected');
      })();

    const cmd = new CommandPacket({
      cmdType: CommandType.REQUEST,
      cmdName,
      data: requestData,
    });
    const payload = cmd.serialize();

    if (this._maxRequestPayloadSize !== null && payload.length > this._maxRequestPayloadSize) {
      throw new PayloadTooLargeError(payload.length, this._maxRequestPayloadSize);
    }

    const sendPayload = this._encryptPayload(payload);
    const containers = s.split(sendPayload);
    for (const c of containers) {
      await this.transport.write(c.serialize());
    }

    this._assembler.reset();
    let firstRead = true;
    for (;;) {
      const notifyData = await this.transport.readNotify(this._readTimeout(firstRead));
      firstRead = false;
      const container = Container.deserialize(notifyData);

      if (container.containerType === ContainerType.CONTROL) {
        this._handleControlError(container);
        continue;
      }

      const result = this._assembler.feed(container);
      if (result !== null) {
        const decrypted = this._decryptPayload(result);
        const resp = CommandPacket.deserialize(decrypted);
        if (resp.cmdType !== CommandType.RESPONSE) {
          throw new ProtocolException(`Expected response, got type=${resp.cmdType}`);
        }
        if (resp.cmdName !== cmdName) {
          throw new ProtocolException(
            `Command name mismatch: expected '${cmdName}', got '${resp.cmdName}'`,
          );
        }
        return resp.data;
      }
    }
  }

  protected async streamReceive(cmdName: string, requestData: Uint8Array): Promise<Uint8Array[]> {
    const s =
      this._splitter ??
      (() => {
        throw new Error('Not connected');
      })();

    const cmd = new CommandPacket({
      cmdType: CommandType.REQUEST,
      cmdName,
      data: requestData,
    });
    const payload = cmd.serialize();

    if (this._maxRequestPayloadSize !== null && payload.length > this._maxRequestPayloadSize) {
      throw new PayloadTooLargeError(payload.length, this._maxRequestPayloadSize);
    }

    const sendPayload = this._encryptPayload(payload);
    const containers = s.split(sendPayload);
    for (const c of containers) {
      await this.transport.write(c.serialize());
    }

    const results: Uint8Array[] = [];
    this._assembler.reset();
    let firstRead = true;
    for (;;) {
      const notifyData = await this.transport.readNotify(this._readTimeout(firstRead));
      firstRead = false;
      const container = Container.deserialize(notifyData);

      if (container.containerType === ContainerType.CONTROL) {
        if (container.controlCmd === ControlCmd.STREAM_END_P2C) break;
        this._handleControlError(container);
        continue;
      }

      const result = this._assembler.feed(container);
      if (result !== null) {
        const decrypted = this._decryptPayload(result);
        const resp = CommandPacket.deserialize(decrypted);
        if (resp.cmdType !== CommandType.RESPONSE) {
          throw new ProtocolException(`Expected response, got type=${resp.cmdType}`);
        }
        results.push(resp.data);
      }
    }
    return results;
  }

  protected async streamSend(
    cmdName: string,
    messages: Uint8Array[],
    finalCmdName: string,
  ): Promise<Uint8Array> {
    const s =
      this._splitter ??
      (() => {
        throw new Error('Not connected');
      })();

    for (const msgData of messages) {
      const cmd = new CommandPacket({
        cmdType: CommandType.REQUEST,
        cmdName,
        data: msgData,
      });
      const payload = cmd.serialize();
      const sendPayload = this._encryptPayload(payload);
      const containers = s.split(sendPayload);
      for (const c of containers) {
        await this.transport.write(c.serialize());
      }
    }

    // Send STREAM_END_C2P
    const tid = s.nextTransactionId();
    const streamEnd = makeStreamEndC2P(tid);
    await this.transport.write(streamEnd.serialize());

    // Wait for final response
    this._assembler.reset();
    let firstRead = true;
    for (;;) {
      const notifyData = await this.transport.readNotify(this._readTimeout(firstRead));
      firstRead = false;
      const container = Container.deserialize(notifyData);

      if (container.containerType === ContainerType.CONTROL) {
        this._handleControlError(container);
        continue;
      }

      const result = this._assembler.feed(container);
      if (result !== null) {
        const decrypted = this._decryptPayload(result);
        const resp = CommandPacket.deserialize(decrypted);
        if (resp.cmdType !== CommandType.RESPONSE) {
          throw new ProtocolException(`Expected response, got type=${resp.cmdType}`);
        }
        if (resp.cmdName !== finalCmdName) {
          throw new ProtocolException(
            `Command name mismatch: expected '${finalCmdName}', got '${resp.cmdName}'`,
          );
        }
        return resp.data;
      }
    }
  }

  disconnect(): void {
    this.transport.disconnect();
    this._session = null;
    this._splitter = null;
  }
}

File: central_flutter/lib/client/blerpc_client.dart
import 'dart:async';
import 'dart:developer' as dev;
import 'dart:typed_data';

import 'package:blerpc_central/proto/blerpc.pb.dart';
import 'package:blerpc_protocol/blerpc_protocol.dart';

import '../ble/ble_transport.dart';
import 'generated_client.dart';

class PayloadTooLargeError implements Exception {
  final int actual;
  final int limit;
  PayloadTooLargeError(this.actual, this.limit);
  @override
  String toString() =>
      'PayloadTooLargeError: Request payload ($actual bytes) exceeds peripheral limit ($limit bytes)';
}

class ResponseTooLargeError implements Exception {
  final String message;
  ResponseTooLargeError(this.message);
  @override
  String toString() => 'ResponseTooLargeError: $message';
}

class PeripheralErrorException implements Exception {
  final int errorCode;
  PeripheralErrorException(this.errorCode);
  @override
  String toString() =>
      'PeripheralErrorException: 0x${errorCode.toRadixString(16).padLeft(2, '0')}';
}

class ProtocolException implements Exception {
  final String message;
  ProtocolException(this.message);
  @override
  String toString() => 'ProtocolException: $message';
}

class BlerpcClient with GeneratedClientMixin {
  final BleTransport transport = BleTransport();
  final bool requireEncryption;

  ContainerSplitter? _splitter;
  final _assembler = ContainerAssembler();
  Duration _timeout = const Duration(milliseconds: 100);
  int? _maxRequestPayloadSize;

  // Encryption state
  BlerpcCryptoSession? _session;

  BlerpcClient({this.requireEncryption = true});

  int get mtu => transport.mtu;
  bool get isEncrypted => _session != null;

  Duration _readTimeout(bool firstRead) {
    if (!firstRead) return _timeout;
    return _timeout > const Duration(seconds: 2)
        ? _timeout
        : const Duration(seconds: 2);
  }

  void _handleControlError(Container container) {
    if (container.controlCmd == ControlCmd.error &&
        container.payload.isNotEmpty) {
      final errorCode = container.payload[0];
      if (errorCode == blerpcErrorResponseTooLarge) {
        throw ResponseTooLargeError(
            "Response exceeds peripheral's max_response_payload_size");
      }
      throw PeripheralErrorException(errorCode);
    }
  }

  Future<List<ScannedDevice>> scan({Duration? timeout}) async {
    return transport.scan(timeout: timeout ?? const Duration(seconds: 5));
  }

  Future<void> connect(ScannedDevice device) async {
    await transport.connect(device);
    _splitter = ContainerSplitter(mtu: transport.mtu);

    try {
      await _requestTimeout();
    } on TimeoutException {
      dev.log('Peripheral did not respond to timeout request, using default');
    }
    try {
      await _requestCapabilities();
    } on TimeoutException {
      dev.log('Peripheral did not respond to capabilities request');
    }

    if (requireEncryption && _session == null) {
      throw StateError(
        'Encryption required but key exchange was not completed. '
        'The peripheral may not support encryption or a MitM may '
        'have stripped the encryption capability flag.',
      );
    }
  }

  Future<void> _requestTimeout() async {
    final s = _splitter!;
    final tid = s.nextTransactionId();
    final req = makeTimeoutRequest(tid);
    await transport.write(req.serialize());
    final data =
        await transport.readNotify(timeout: const Duration(seconds: 1));
    final resp = Container.deserialize(data);
    if (resp.containerType == ContainerType.control &&
        resp.controlCmd == ControlCmd.timeout &&
        resp.payload.length == 2) {
      final bd = ByteData.sublistView(resp.payload);
      final ms = bd.getUint16(0, Endian.little);
      _timeout = Duration(milliseconds: ms);
      dev.log('Peripheral timeout: ${ms}ms');
    }
  }

  Future<void> _requestCapabilities() async {
    final s = _splitter!;
    final tid = s.nextTransactionId();
    final req = makeCapabilitiesRequest(tid);
    await transport.write(req.serialize());
    final data =
        await transport.readNotify(timeout: const Duration(seconds: 1));
    final resp = Container.deserialize(data);
    if (resp.containerType == ContainerType.control &&
        resp.controlCmd == ControlCmd.capabilities &&
        resp.payload.length >= 6) {
      final bd = ByteData.sublistView(resp.payload);
      final maxReq = bd.getUint16(0, Endian.little);
      final maxResp = bd.getUint16(2, Endian.little);
      final flags = bd.getUint16(4, Endian.little);
      _maxRequestPayloadSize = maxReq;
      dev.log(
          'Capabilities: max_req=$maxReq, max_resp=$maxResp, flags=0x${flags.toRadixString(16).padLeft(4, '0')}');

      if (flags & capabilityFlagEncryptionSupported != 0) {
        await _performKeyExchange();
      }
    }
  }

  Future<void> _performKeyExchange() async {
    final s = _splitter!;

    try {
      _session = await centralPerformKeyExchange(
        send: (payload) async {
          final tid = s.nextTransactionId();
          final req = makeKeyExchange(tid, payload);
          await transport.write(req.serialize());
        },
        receive: () async {
          final data =
              await transport.readNotify(timeout: const Duration(seconds: 2));
          final resp = Container.deserialize(data);
          if (resp.containerType != ContainerType.control ||
              resp.controlCmd != ControlCmd.keyExchange) {
            throw StateError(
                'Expected KEY_EXCHANGE response, got something else');
          }
          return resp.payload;
        },
      );
      dev.log('E2E encryption established');
    } catch (e) {
      dev.log('Key exchange failed: $e');
      if (requireEncryption) rethrow;
    }
  }

  Future<Uint8List> _encryptPayload(Uint8List payload) async {
    if (_session == null) {
      if (requireEncryption) {
        throw StateError('Encryption required but no session established');
      }
      return payload;
    }
    return _session!.encrypt(payload);
  }

  Future<Uint8List> _decryptPayload(Uint8List payload) async {
    if (_session == null) {
      if (requireEncryption) {
        throw StateError('Encryption required but no session established');
      }
      return payload;
    }
    return _session!.decrypt(payload);
  }

  @override
  Future<Uint8List> call(String cmdName, Uint8List requestData) async {
    final s = _splitter ?? (throw StateError('Not connected'));

    final cmd = CommandPacket(
      cmdType: CommandType.request,
      cmdName: cmdName,
      data: requestData,
    );
    final payload = cmd.serialize();

    if (_maxRequestPayloadSize != null &&
        payload.length > _maxRequestPayloadSize!) {
      throw PayloadTooLargeError(payload.length, _maxRequestPayloadSize!);
    }

    final sendPayload = await _encryptPayload(payload);
    final containers = s.split(sendPayload);
    for (final c in containers) {
      await transport.write(c.serialize());
    }

    _assembler.reset();
    var firstRead = true;
    while (true) {
      final notifyData =
          await transport.readNotify(timeout: _readTimeout(firstRead));
      firstRead = false;
      final container = Container.deserialize(notifyData);

      if (container.containerType == ContainerType.control) {
        _handleControlError(container);
        continue;
      }

      final result = _assembler.feed(container);
      if (result != null) {
        final decrypted = await _decryptPayload(result);
        final resp = CommandPacket.deserialize(decrypted);
        if (resp.cmdType != CommandType.response) {
          throw ProtocolException(
              'Expected response, got type=${resp.cmdType}');
        }
        if (resp.cmdName != cmdName) {
          throw ProtocolException(
              "Command name mismatch: expected '$cmdName', got '${resp.cmdName}'");
        }
        return resp.data;
      }
    }
  }

  @override
  Future<List<Uint8List>> streamReceive(
      String cmdName, Uint8List requestData) async {
    final s = _splitter ?? (throw StateError('Not connected'));

    final cmd = CommandPacket(
      cmdType: CommandType.request,
      cmdName: cmdName,
      data: requestData,
    );
    final payload = cmd.serialize();

    if (_maxRequestPayloadSize != null &&
        payload.length > _maxRequestPayloadSize!) {
      throw PayloadTooLargeError(payload.length, _maxRequestPayloadSize!);
    }

    final sendPayload = await _encryptPayload(payload);
    final containers = s.split(sendPayload);
    for (final c in containers) {
      await transport.write(c.serialize());
    }

    final results = <Uint8List>[];
    _assembler.reset();
    var firstRead = true;
    while (true) {
      final notifyData =
          await transport.readNotify(timeout: _readTimeout(firstRead));
      firstRead = false;
      final container = Container.deserialize(notifyData);

      if (container.containerType == ContainerType.control) {
        if (container.controlCmd == ControlCmd.streamEndP2C) break;
        _handleControlError(container);
        continue;
      }

      final result = _assembler.feed(container);
      if (result != null) {
        final decrypted = await _decryptPayload(result);
        final resp = CommandPacket.deserialize(decrypted);
        if (resp.cmdType != CommandType.response) {
          throw ProtocolException(
              'Expected response, got type=${resp.cmdType}');
        }
        results.add(resp.data);
      }
    }
    return results;
  }

  @override
  Future<Uint8List> streamSend(
      String cmdName, List<Uint8List> messages, String finalCmdName) async {
    final s = _splitter ?? (throw StateError('Not connected'));

    for (final msgData in messages) {
      final cmd = CommandPacket(
        cmdType: CommandType.request,
        cmdName: cmdName,
        data: msgData,
      );
      final payload = cmd.serialize();
      final sendPayload = await _encryptPayload(payload);
      final containers = s.split(sendPayload);
      for (final c in containers) {
        await transport.write(c.serialize());
      }
    }

    // Send STREAM_END_C2P
    final tid = s.nextTransactionId();
    final streamEnd = makeStreamEndC2P(tid);
    await transport.write(streamEnd.serialize());

    // Wait for final response
    _assembler.reset();
    var firstRead = true;
    while (true) {
      final notifyData =
          await transport.readNotify(timeout: _readTimeout(firstRead));
      firstRead = false;
      final container = Container.deserialize(notifyData);

      if (container.containerType == ContainerType.control) {
        _handleControlError(container);
        continue;
      }

      final result = _assembler.feed(container);
      if (result != null) {
        final decrypted = await _decryptPayload(result);
        final resp = CommandPacket.deserialize(decrypted);
        if (resp.cmdType != CommandType.response) {
          throw ProtocolException(
              'Expected response, got type=${resp.cmdType}');
        }
        if (resp.cmdName != finalCmdName) {
          throw ProtocolException(
              "Command name mismatch: expected '$finalCmdName', got '${resp.cmdName}'");
        }
        return resp.data;
      }
    }
  }

  // Streaming convenience methods (manually implemented, not auto-generated)

  Future<List<(int, int)>> counterStreamAll(int count) async {
    final req = CounterStreamRequest()..count = count;
    final responses = await streamReceive(
        'counter_stream', Uint8List.fromList(req.writeToBuffer()));
    return responses.map((data) {
      final resp = CounterStreamResponse.fromBuffer(data);
      return (resp.seq, resp.value);
    }).toList();
  }

  Future<CounterUploadResponse> counterUploadAll(int count) async {
    final messages = List.generate(count, (i) {
      final req = CounterUploadRequest()
        ..seq = i
        ..value = i * 10;
      return Uint8List.fromList(req.writeToBuffer());
    });
    final respData =
        await streamSend('counter_upload', messages, 'counter_upload');
    return CounterUploadResponse.fromBuffer(respData);
  }

  void disconnect() {
    transport.disconnect();
    _session = null;
    _splitter = null;
  }
}
